compiler/hash-exhaustiveness/src/lib.rs

```diff
@@ -1 +1,273 @@
+//! for example `Some(4)`. Therefore, the exhaustiveness check will conclude
+//! that the provided pattern vector is not exhaustive and misses some cases.
+//!
+//! The implementation of this algorithm is based on the research paper:
+//!
+//! <http://moscova.inria.fr/~maranget/papers/warn/warn.pdf>
+//!
+//! inspired by and based off of the Rust Compiler implementation:
+//!
+//! <https://github.com/rust-lang/rust/tree/master/compiler/rustc_mir_build/src/thir/pattern/usefulness.rs>
+#![feature(let_chains, if_let_guard)]
 
+pub mod constant;
+pub mod construct;
+pub mod deconstruct;
+pub mod diagnostics;
+pub mod fields;
+pub mod list;
+pub mod lower;
+pub mod matrix;
+pub mod range;
+pub mod stack;
+pub mod storage;
+pub mod usefulness;
+pub mod wildcard;
+
+use construct::DeconstructedCtor;
+use deconstruct::DeconstructedPat;
+use diagnostics::{ExhaustivenessDiagnostics, ExhaustivenessError, ExhaustivenessWarning};
+use hash_ast::ast::MatchOrigin;
+use hash_intrinsics::primitives::{AccessToPrimitives, DefinedPrimitives};
+use hash_reporting::diagnostic::Diagnostics;
+use hash_source::location::SourceLocation;
+use hash_tir::{
+    environment::env::{AccessToEnv, Env},
+    pats::PatId,
+    tys::TyId,
+};
+use hash_utils::store::CloneStore;
+use storage::{
+    DeconstructedCtorId, DeconstructedCtorStore, DeconstructedPatId, DeconstructedPatStore,
+    ExhaustivenessCtx,
+};
+use usefulness::Reachability;
+
+/// General exhaustiveness context that's used when performing
+/// splitting and specialisation operations.
+#[derive(Copy, Clone)]
+pub struct PatCtx {
+    /// The term of the current column that is under investigation
+    pub ty: TyId,
+    /// Whether the current pattern is the whole pattern as found in a match
+    /// arm, or if it's a sub-pattern.
+    pub(crate) is_top_level: bool,
+}
+
+impl PatCtx {
+    /// Create a new [PatCtx]
+    pub fn new(ty: TyId, is_top_level: bool) -> Self {
+        PatCtx { ty, is_top_level }
+    }
+}
+
+pub struct ExhaustivenessChecker<'tc> {
+    /// The span of the subject that is being checked for exhaustiveness
+    /// or usefulness.
+    subject_span: SourceLocation,
+
+    /// A reference to the TC env in order to lookup various TIR items and have
+    /// access to TC utilities.
+    tcx: &'tc Env<'tc>,
+
+    /// The primitive definitions that are needed for creating and comparing
+    /// primitive types with the TIR.
+    primitives: &'tc DefinedPrimitives,
+
+    /// The exhaustiveness store keeps track of created de-constructed patterns
+    /// and usefulness constructors, basically representations used by
+    /// exhaustiveness to evaluate whether something is exhaustive. Since
+    /// every exhaustive instance is independent from one another, we can keep
+    /// the [ExhaustivenessCtx] local to each checker, or in other words, local
+    /// to the thread that is processing the job.
+    ecx: ExhaustivenessCtx,
```

Comment:
Can exhaustiveness stores become thread local static?

Reply:
They can but I don’t see any need to share them across several exhaustiveness instances. The idea is in the future when they run on their own threads, they are independent and don’t interfere with one another.

Reply:
Yeah that's what I mean, thread locals. Just so that you don't need to carry them around in the context.

Reply:
Ah yes, they can. Ideally, they should be cleared after TC so they don't take up excess memory. I guess this can be done in the `cleanup()` call in the `SemanticAnalysis`? Can I make the stores thread local by just using `thread_local!` and use the same technique as in TIR?

Reply:
Take a look at `hash_tir:: environment::stores`. Probably have to do something similar. A lot of it might be able to be factored out into `hash_utils`.

Reply:
Decided that we'll do this later and migrate exhaustiveness when its ready.